choral/src/main/java/choral/compiler/amend/RelaxedTyper.java

```diff
@@ -0,0 +1,2828 @@
+
+		public CallableScope( ClassOrInterfaceScope parent, Member.HigherCallable callable ) {
+			this.parent = parent;
+			this.callable = callable;
+		}
+
+		private final ClassOrInterfaceScope parent;
+
+		@Override
+		protected ClassOrInterfaceScope parent() {
+			return parent;
+		}
+
+		@Override
+		public Optional< ? extends HigherDataType > lookupDataType( String query ) {
+			Optional< ? extends HigherDataType > result = callable.typeParameter( query );
+			if( result.isEmpty() ) {
+				return super.lookupDataType( query );
+			}
+			return result;
+		}
+
+		@Override
+		public Optional< ? extends HigherReferenceType > lookupReferenceType( String query ) {
+			Optional< ? extends HigherReferenceType > result = callable.typeParameter( query );
+			if( result.isEmpty() ) {
+				return super.lookupReferenceType( query );
+			}
+			return result;
+		}
+
+		@Override
+		public Optional< ? extends HigherTypeParameter > lookupTypeParameter( String query ) {
+			Optional< ? extends HigherTypeParameter > result = callable.typeParameter( query );
+			if( result.isEmpty() && !callable.isStatic() ) {
+				return super.lookupTypeParameter( query );
+			}
+			return result;
+		}
+
+		private final Map< HigherTypeParameter, TypeParameterScope > typeParameterScopes = new HashMap<>();
+
+		@Override
+		public TypeParameterScope getScope( HigherTypeParameter parameter ) {
+			TypeParameterScope scope = typeParameterScopes.get( parameter );
+			if( scope == null ) {
+				scope = new TypeParameterScope( this, parameter );
+				typeParameterScopes.put( parameter, scope );
+			}
+			return scope;
+		}
+
+		private CallableBodyScope bodyScope;
+
+		public CallableBodyScope getScope() {
+			if( bodyScope == null ) {
+				bodyScope = new CallableBodyScope( this );
+			}
+			return bodyScope;
+		}
+	}
+
+	private static final class CallableBodyScope extends ChildScope
+			implements VariableDeclarationScope {
+
+		private CallableBodyScope( CallableScope parent ) {
+			this.parent = parent;
+			for( Signature.Parameter p : parent.callable.innerCallable().signature().parameters() ) {
+				variables.put( p.identifier(), p.type() );
+			}
+		}
+
+		private final CallableScope parent;
+
+		@Override
+		public CallableScope parent() {
+			return parent;
+		}
+
+		private final Map< String, GroundDataType > variables = new HashMap<>();
+
+		@Override
+		public List<GroundDataType> getChannels(){
+			List<GroundDataType> channels = new ArrayList<>();
+			HigherDataType diDataChannel = assertLookupDataType("choral.channels.DiDataChannel");
+			HigherDataType diSelectChannel = assertLookupDataType("choral.channels.DiSelectChannel");
+			
+
+			// this fields
+			lookupThis().fields().forEach( field -> {
+				
+				if( field.type().typeConstructor() instanceof HigherInterface ){
+					HigherInterface typec = (HigherInterface)field.type().typeConstructor();
+					if( typec.isSubtypeOf_relaxed( diDataChannel ) || typec.isSubtypeOf_relaxed( diSelectChannel ) ){
```

Comment:
yes, I was planning to figure out which channels could be used for each inference individually, but I might change this later, when I have a better idea of how exactly I want to channels. I am considering that it might be nice to have, instead of just a list of channels, a map, mapping worlds to a list of channels they can send through. Maybe I want two lists, one for data channels and one for select channels with channels that can act as both appearing in both lists. 

How I store channels, and how I sort them will probably change when I get into inference more